corgidrp/calibrate_kgain.py

```diff
@@ -0,0 +1,1099 @@
+        when k-gain is the primary desired product, since it is known more 
+        accurately than non-unity values. stack_arr and stack_arr2 must be 
+        obtained under the same positioning of the pupil relative to the 
+        detector.
+    
+    emgain : float
+        The value of the measured/actual EM gain used to collect the frames used 
+        to build the stack_arr and stack_arr2 arrays. Must be >= 1.0. (note: 
+        unity em gain is recommended when k-gain is the primary desired product, 
+        since it is known more accurately than non-unity values.)
+    
+    min_val : int
+        Minimum value (in DN) of mean values from sub-stacks to use in calculating 
+        kgain. (> 400 recommended)
+        
+    max_val : int
+        Maximum value (in DN) of mean values from sub-stacks to use in calculating 
+        kgain. Choose value that avoids large deviations from linearity at high 
+        counts. (< 6,000 recommended)
+    
+    binwidth : int
+        Width of each bin for calculating std devs and means from each 
+        sub-stack in stack_arr. Maximum value of binwidth is 800. NOTE: small 
+        values increase computation time.
+        (minimum 10; binwidth between 65 and 75 recommended)
+    
+    mkplot : boolean
+        Option to display plots. Default is None. If mkplot is anything other 
+        than None, then this option is chosen.
+        
+    verbose : boolean
+        Option to display various diagnostic print messages. Default is None. 
+        If mkplot is anything other than None, then this option is chosen.
+    
+    Returns
+    -------
+    kgain: KGain data type
+        kgain estimate from the least-squares fit to the photon transfer curve 
+        (in e-/DN). The expected value of kgain for EXCAM with flight readout 
+        sequence should be between 8 and 9 e-/DN.
+    
+    read_noise_gauss : float
+        Read noise estimate from the prescan regions (in e-), calculated from 
+        the Gaussian fit std devs (in DN) multiplied by kgain. This value 
+        should be considered the true read noise, not affected by the fixed 
+        pattern noise. 
+    
+    read_noise_stdev : float
+        Read noise estimate from the prescan regions (in e-), calculated from 
+        simple std devs (in DN) multiplied by kgain. This value should be 
+        larger than read_noise_gauss and is affected by the fixed pattern noise.
+    
+    ptc : array-like
+        array of size N x 2, where N is the number of bins set by the 'signal_bins_N' 
+        parameter in detector.kgain_params. The first column is the mean (DN) and 
+        the second column is standard deviation (DN) corrected for read noise.
+    
+    """
+    # copy stack_arr and stack_arr2 and cast them into np arrays for convenience
+    stack_arr, stack_arr2 = copy_and_cast(stack_arr, stack_arr2)
+
+    # input checks
+    from corgidrp.detector import kgain_params as master_files
```

Comment:
I think it would be better to put this line at the very top.  (Side note:  The implementation of input parameters will be overhauled in Jason's PR #106 (away from loading them from detector.py), but I think this is fine as it is, and that PR would handle any changes needed.)

Reply:
Update: Since the new implementation of the DetectorParams class has been pulled in, I think these parameters should go in there.  Then you would have a class instance of that as an input to your function to get these k gain parameters.

Reply:
Are some (or all) of the parameters just better as input kwargs? I haven't thought through the pros/cons, but I thought I'd throw it out there for discussion.

Reply:
I guess that could work if the walker can handle kwargs.  I just suggested the move to DetectorParams so that all input parameters are in one place.  On that note, I was going through my list of things to do, and one of them was to suggest putting all the calibration function and step function input parameter arguments in DetectorParams so that they would all be in one place, but if you and Jason prefer to keep parameters as kwargs associated with their relevant functions (and if the walker can handle that), I suppose that's fine.  As long as the CTC can set the kwargs and DetectorParams as needed when this is in Docker, I am fine either way.

Reply:
This conversation has been moved over to Issue #126 for a more general discussion (thanks @kjl0025 for starting it there!).

Reply:
default of the boolean mkplot should be False. Do we want matplotlib in the requirement.txt?

Reply:
I think @maxwellmb prefers plots to be stored. Right now, default is None, which is probably fine. Should we leave None as default in the calibration functions? And let the ut do the plots and store them? If so, "Do we want matplotlib in the requirement.txt?"

Reply:
I sent an email to Guillermo Gonzalez (tellus1) who wrote these functions, and CC'ed Max.

Reply:
Added matplotlib to requirements.txt and order packages alphabetically

Reply:
> Yes, let's add matplotlib to the requirement.txt!
> 
> Looking at the `kgain_params`, I don't know exactly what all of them are, but here's my guess as to where different ones should go:
> 
> * All of the row and column ones _might_ be constants, in which case they could stay in a `kgain_params` dictionary, but I'd move the location of that dictionary to this file rather than detector.py since I don't think they're used for anything else. They might also just be fine as kwargs here.
> * The rest of them are probably best as kwargs, since we'll likely want to be able to change them around with real data.

@maxwellmb When using kwargs, do you mean to just type their keys/values in the unit tests directly?

Reply:
I think the answer is in the next thread

Reply:
Yep!

Reply:
Done